otherlibs/stdune/fpath.ml

```diff
@@ -15,60 +14,35 @@ let mkdir ?(perms = 0o777) t_s =
   with
   | Unix.Unix_error (EEXIST, _, _) -> Already_exists
   | Unix.Unix_error (ENOENT, _, _) -> Missing_parent_directory
-  | Unix.Unix_error (ENOTDIR, _, _) -> Parent_not_directory
 
 type mkdir_p_result =
   | Already_exists
   | Created
-  | Already_exists_not_directory of string
-
-let dyn_of_mkdir_p_result = function
-  | Already_exists -> Dyn.Variant ("Already_exists", [])
-  | Already_exists_not_directory path ->
-    Dyn.Variant ("Already_exists_not_directory", [ Dyn.string path ])
-  | Created -> Dyn.Variant ("Created", [])
 
 let rec mkdir_p ?perms t_s =
   match mkdir ?perms t_s with
   | Created -> Created
-  | Already_exists ->
-    if Stdlib.Sys.is_directory t_s then Already_exists
-    else Already_exists_not_directory t_s
+  | Already_exists -> Already_exists
   | Missing_parent_directory -> (
     if is_root t_s then
       Code_error.raise
         "Impossible happened: [Fpath.mkdir] refused to create a directory at \
-         the root, allegedly because its parent was missing."
+         the root, allegedly because its parent was missing"
         []
     else
       let parent = Filename.dirname t_s in
       match mkdir_p ?perms parent with
-      | Already_exists_not_directory _ -> Already_exists_not_directory parent
       | Created | Already_exists -> (
         (* The [Already_exists] case might happen if some other process managed
            to create the parent directory concurrently. *)
         match mkdir t_s ?perms with
         | Created -> Created
         | Already_exists -> Already_exists
-        | Missing_parent_directory | Parent_not_directory ->
+        | Missing_parent_directory ->
           (* But we just successfully created the parent directory. So it was
-             likely deleted or overwritten right now. Let's give up *)
-          Code_error.raise "Failed to create parent directory."
+             likely deleted right now. Let's give up *)
+          Code_error.raise "failed to create parent directory"
             [ ("t_s", Dyn.string t_s) ]))
-  | Parent_not_directory ->
```

Comment:
I think this is the only valid part of this PR. We should only try to improve the error message when we know we're going to fail anyway.